addons/sale_loyalty/models/sale_order.py

```diff
@@ -997,7 +1000,8 @@ def _try_apply_code(self, code):
 
         if not program or not program.active:
             return {'error': _('This code is invalid (%s).', code), 'not_found': True}
-        elif (program.limit_usage and program.total_order_count >= program.max_usage):
+        elif (program.limit_usage and program.total_order_count >= program.max_usage) or\
+            (program.date_to and program.date_to < fields.Date.context_today(self)):
```

Comment:
I hate the \ in code, they are unclear IMHO
```suggestion
        elif (
            (program.limit_usage and program.total_order_count >= program.max_usage))
            or (program.date_to and program.date_to < fields.Date.context_today(self))
        ):
```